### PR TITLE
translate alias of named symbols

### PR DESCRIPTION
**Pull request by milahu**
support tree-sitter grammars like

```js
module.exports = grammar({
  rules: {
    a: $ => "aaa",
    c: $ => alias($.a, $.b), // alias of the named symbol a
  }
}
```

tested on [tree-sitter-make](https://github.com/alemuller/tree-sitter-make) ([result](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/tree/master/test/cases/make/lezer-parser-make/src))

not sure if the tokens-logic is needed (`this.tokenNewNames` etc)

### CODE REVIEW

File: src/import.ts [hunk 1]
```diff
@@ -96,8 +99,17 @@ class Context {
...
         return opt ? this.translateInner(opt, token, 10) + "?"
           : expr.members.map(e => this.translateInner(e, token, prec(expr))).join(" | ")
-      case "ALIAS": // FIXME this should override/drop the name of the inner expr, somehow
-        if (token) throw new RangeError("Alias expression in token")
+      case "ALIAS":
+        if (expr.content.type == "SYMBOL" && expr.content.name[0] != "_") {
+          // override the name of the inner symbol
+          let hiddenName = this.hiddenSymbols[expr.content.name]
+          if (!hiddenName) {
+            hiddenName = this.generateName("_" + expr.content.name)
```
Comment (milahu) [inline src/import.ts:107]: handle name collisions = when `("_" + expr.content.name)` exists

**Comment by marijnh** [comment]
I do not understand what this does. Can you give an example of what kind of output you are producing for a renamed rule?

**Comment by milahu** [comment]
the inner symbol is hidden by adding a underscore-prefix "_" to the tree-sitter symbol name

i added a test case in my fork: [test-alias-of-named-symbol](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/tree/11ab876d314a1e3fd41e4145c9b28d85e045ac1f/test/cases/test-alias-of-named-symbol)

[tree-sitter grammar](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/blob/11ab876d314a1e3fd41e4145c9b28d85e045ac1f/test/cases/test-alias-of-named-symbol/tree-sitter-test-alias-of-named-symbol/grammar.js)

```js
module.exports = grammar({
  name: "test_alias_of_named_symbol",

  rules: {
    document: $ => repeat(choice($.word, $.parens_word)),
    word: $ => /[a-z]+/,
    parens_word: $ => seq("(", alias($.word, $.word_in_parens), ")"),
  },
});
```

[tree-sitter test](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/blob/11ab876d314a1e3fd41e4145c9b28d85e045ac1f/test/cases/test-alias-of-named-symbol/tree-sitter-test-alias-of-named-symbol/test/corpus/hello.txt)

```
================================================================================
hello node
================================================================================

a (b) c

--------------------------------------------------------------------------------

(document
  (word)
  (parens_word
    (word_in_parens))
  (word))
```



[lezer grammar](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/blob/11ab876d314a1e3fd41e4145c9b28d85e045ac1f/test/cases/test-alias-of-named-symbol/lezer-parser-test-alias-of-named-symbol/src/grammar.lezer)

```
@top Document {
  (Word | ParensWord)*
}

ParensWord {
  "(" WordInParens { word1 } ")"
}

@skip { token_1 }

@tokens {
  token_1 {
    $[ \t\r\n]
  }
  Word { word1 }
  word1 {
    $[a-z]+
  }
}
```

[lezer test](https://github.com/milahu/lezer-parser-import-tree-sitter-scanner/blob/8102a4d960b88703195c650c2ad1e0ccfa77b33e/test/cases/test-alias-of-named-symbol/lezer-parser-test-alias-of-named-symbol/test/hello.txt)

```
# test
a (b) c
==>
Document (
  Word,
  ParensWord (
    WordInParens
  ),
  Word
)
```

> not sure if the tokens-logic is needed

yes, see test: Word is a token

**Comment by marijnh** [comment]
I see, it renames the original token in order to be able to wrap it. I can see this introducing new issues (of ambiguous tokens), and my approach so far was that this kind of renaming or wrapping is something the user will have to address manually.

Again, I don't want to spend much time maintaining this, so I don't want to complicate it with further features or spend time reviewing patches. If you want to move this tool forward, I suggest you work on a fork instead. I'm going to decline this patch.

**Pull request closed without merging**